### DISCUSSION TITLE
Add html Webpage from data folder in SPIFF & run on APmode when HOME is pressed.

### DISCUSSION THREAD

**Discussion by xtech007** [Hieromon/AutoConnect#334] [Q&A] [answered]
Hi all!
I’m really new and super amazed of all this Library can do!  Just started to experiment following all the superb documentation provided to make modifications to customize its menu items.

As is, I have a project where it’s a standalone webpage served from the ESP32 webserver and WiFi. So to connect is simple; turn on the esp32, on the user device WiFi settings look for the esp32 WiFi network, connect to it, add password, once connected you are in the webpage!

My question would be, How can I add my webpage that sits on the SPIFF under a folder named data/page1.html And Run when the HOME button is pressed in the AutoConnect Menu?

I have read the Custom page implementation but as you can see my webpage is pure html not json.

**Comment by Hieromon** [comment, marked as answer]
A description of that technique is given in the documentation as [Capture the legacy web pages as items into the menu](https://hieromon.github.io/AutoConnect/adexterior.html#capture-the-legacy-web-pages-as-items-into-the-menu).
The following sketch coded in the above section is a WebServer native page handler:

```cpp
server.on("/hello", [](){
  server.send(200, "text/html", String(F(
"<html>"
"<head><meta name='viewport' content='width=device-width,initial-scale=1.0'></head>"
"<body><h2>Hello, world</h2></body>"
"</html>"
  )));
});

portal.append("/hello", "HELLO");
```

In that example, it is added as a "HELLO" item into the menu, but you can add a legacy page handler to the **HOME** menu item by using the [AutoConnectConfig::homeUri](https://hieromon.github.io/AutoConnect/apiconfig.html#homeuri) setting.

So, what you need may be had in the sketch below:

```cpp
#include <Arduino.h>
#include <WiFi.h>
#include <WebServer.h>
#include <SPIFFS.h>
#include <AutoConnect.h>

// Change HOME URI and HTML file path you wish.
#define HOMEURI       "/home"
#define ADDONPAGEFILE "/data/page1.html"

WebServer         server;
AutoConnect       portal(server);
AutoConnectConfig config;

bool loadAddonPage(const char* path) {
  bool  rc = false;
  Serial.printf("Page %s requested, ", path);
  File  addonPage = SPIFFS.open(path, "r");
  if (addonPage) {
    if (!addonPage.isDirectory()) {
      server.streamFile(addonPage, "text/html");
      Serial.println("successfully loaded");
      rc = true;
    }
    else
      Serial.println("it's directory");
    addonPage.close();
  }
  else
    Serial.println("open failed, probably does not exist.");
  return rc;
}

void setup() {
  delay(1000);
  Serial.begin(115200);
  Serial.println();

  server.on(HOMEURI, HTTP_GET, []() {
    if (!loadAddonPage(ADDONPAGEFILE))
      server.send(404, "text/plain", "File not found");
  });

  SPIFFS.begin(true);
  config.homeUri = HOMEURI;   // It's important
  portal.config(config);      // Register the HOMEPATH to HOME item of the menu
  portal.begin();
}

void loop() {
  portal.handleClient();
}
```

**Reply by xtech007** [reply]
Thank you for your fast response and your time!
I uploaded your sketch and gave me an error when I pressed on the home button on the Portal. “File not found”.

Maybe related to the SPIFF, the way I uploaded the data folder containing the page1.html is using the esp32 sketch data upload tool.

**Reply by xtech007** [reply]
<img alt="409FD1D9-75C1-4A22-AF3E-7F29AEB344A3" src="https://user-images.githubusercontent.com/49664309/112414565-a00e4580-8cf8-11eb-9c69-1bbad57bc96e.jpeg" width="480px">


<img alt="E7BC3B37-5517-4E79-A389-5A5A5086F44F" src="https://user-images.githubusercontent.com/49664309/112414567-a0a6dc00-8cf8-11eb-8170-b5790b15d72c.jpeg" width="480px">


EDIT by Hieromon
Since the display size for a resolution is too high of the original PGN image becomes excessive, it was the proper size by changing the original post.

**Comment by xtech007** [comment]
And on the terminal this is what it shows:

<img alt="B120AD3B-6BE3-4A4E-B6F9-1158ACE4CBEA" src="https://user-images.githubusercontent.com/49664309/112415380-1cedef00-8cfa-11eb-9820-6bea3ce6939f.jpeg" width="480px">

**EDIT** by Hieromon
Since the display size for a resolution is too high of the original PGN image becomes excessive, it was the proper size by changing the original post.

**Reply by Hieromon** [reply]
The sketch I uploaded above has been tested and should work.
I was a little worried about the location of the `page1.html` file you specified. The filesystem uploader to SPIFFS is usually containerized in the `data` subfolder where the sketch resides. The `data` subfolder is the uploader's enclosure and its location is the **root of SPIFFS**.
Therefore, in the case of **the named data/page1.html on the SPIFFS**, the folder structure that the file system uploader of Arduino IDE containers should be following:

```
<Sketch folder>
 |
 +-<data>
    |
    +-<data>
      |
      +- page1.html
```

If your intention specify to `/page1.html` on SPIFFS, change the define directive as `#define ADDONPAGEFILE "/data/page1.html"` to `#define ADDONPAGEFILE "/page1.html"`.

**Reply by xtech007** [reply]
Thank you so Much for your support!
I loaded the new modified #define ADDONPAGEFILE "/page1.html part and Worked!!
But, I can not really see the original website ,it's like if all the style and script where removed from it.

**Reply by Hieromon** [reply]
It is under the captive portal opened by your cellphone OS. In that state, the ESP module cannot connect to the Internet. If your HTML contains other sites or URIs, all those links will be invalid.
Also, You are connecting to the AP side of the ESP32, 172.217.28.1.
If the ESP32 module has a connection to the AP, exit from the captive portal once then access the ESP32 STA address from your browser.

**Reply by xtech007** [reply]
Oh, Ok!
So I would only reach my page when connected to my router or internet not in AP mode?

**Comment by xtech007** [comment]
This is what it’s supposed to look like!
<img alt="57C3A300-0E55-4E37-AD61-3B94401B4EA1" src="https://user-images.githubusercontent.com/49664309/112574550-e7620800-8dc4-11eb-8a33-b4bf4688ed73.jpeg" width="360px">

**EDIT** by Hieromon
Since the display size for a resolution is too high of the original PGN image becomes excessive, it was the proper size by changing the original post.

**Comment by xtech007** [comment]
And this is what I get when press the home button in the portal.
<img alt="E8DF9B7A-9524-4E7E-A740-2E1C75114D33" src="https://user-images.githubusercontent.com/49664309/112574607-0b254e00-8dc5-11eb-9bab-9e15aab170df.png" width="360px">

**EDIT** by Hieromon
Since the display size for a resolution is too high of the original PGN image becomes excessive, it was the proper size by changing the original post.

**Comment by xtech007** [comment]
The links don’t work and the images don’t rendered.

**Reply by Hieromon** [reply]
I had modified your posts above to show the images correctly.

**Comment by xtech007** [comment]
How can I reach out to you, privately?
I have a project in mind, and would need something like the portal with some personal modifications. I don’t have the technical background to come near of what you have accomplished.
So why re-invent the wheel?
Folks at Freelancer and upwork claim to be able to accomplish the task but they fail and after all it’s your project. They wants some serious $$ for your work.  

I rather contribute to the person or group that developed the portal.
I’m confident the requirements would be met and most likely 💯% working with your library.

**Reply by xtech007** [reply]
By the way the links, are actually other pages (page2.html -page4.html) in the same spiff. I can send them for your review if interested. If so my Email is Leonel.facundo@yahoo.com.

**Reply by Hieromon** [reply]
Unfortunately, it is clear that the content of page1.htlm has a stylesheet that depends on the outside of the ESP32 host. (Maybe that CSS link is also present in page2.html, page3.html, page4.html). Perhaps their content also hides tricks that were shrunk by jQuery. It may also be a mix of Web APIs provided by other websites. In any case, they do not contribute to the improvement of the AutoConnect library.
So that's why I'm not interested in your HTML. It is for you to comprehend their meaning with your own. I hope the Discussions is a place where we can provide useful information to other users of AutoConnect.

Thank you.

**Reply by xtech007** [reply]
Thank you again for your time!
I have done everything I can possibly think and still can’t get the results I hoped for.

Can I hire you to makes the proper changes and have it work?
